gossip/src/protocol.rs

```diff
@@ -81,13 +78,108 @@ pub(crate) struct PruneData {
 }
 
 impl Protocol {
+    // Implement bincode::serialize_into for Protocol.
+    pub(crate) fn bincode_serialize<W: Write>(&self, writer: &mut W) -> Result<(), bincode::Error> {
+        match self {
+            Self::PullRequest(filter, caller) => {
+                writer.write_all(&0u32.to_le_bytes())?;
+                bincode::serialize_into(&mut *writer, filter)?;
+                Ok(caller.bincode_serialize(writer)?)
+            }
+            Self::PullResponse(pubkey, values) => {
+                writer.write_all(&1u32.to_le_bytes())?;
+                writer.write_all(pubkey.as_ref())?;
+                Ok(CrdsValue::bincode_serialize_many(values, writer)?)
+            }
+            Self::PushMessage(pubkey, values) => {
+                writer.write_all(&2u32.to_le_bytes())?;
+                writer.write_all(pubkey.as_ref())?;
+                Ok(CrdsValue::bincode_serialize_many(values, writer)?)
+            }
+            Self::PruneMessage(pubkey, prune_data) => {
+                writer.write_all(&3u32.to_le_bytes())?;
+                writer.write_all(pubkey.as_ref())?;
+                bincode::serialize_into(writer, prune_data)
+            }
+            Self::PingMessage(ping) => {
+                writer.write_all(&4u32.to_le_bytes())?;
+                bincode::serialize_into(writer, ping)
+            }
+            Self::PongMessage(pong) => {
+                writer.write_all(&5u32.to_le_bytes())?;
+                bincode::serialize_into(writer, pong)
+            }
+        }
+    }
+
+    // Implement bincode::deserialize_from for Protocol.
+    pub(crate) fn bincode_deserialize(bytes: &[u8]) -> Result<Self, bincode::Error> {
+        fn bincode_deserialize<T: DeserializeOwned>(bytes: &[u8]) -> Result<T, bincode::Error> {
+            bincode::options()
+                .with_limit(bytes.len() as u64)
+                .with_fixint_encoding()
+                .reject_trailing_bytes()
+                .deserialize(bytes)
+        }
+        let (tag, bytes) = convert_fixed_bytes::<[u8; 4], 4>(bytes)?;
```

Comment:
same here, can you elaborate on what the `4s` are here?

Reply:
oh so get the first 4 bytes of the serialized value and convert that into a "tag" which just represents the type of message received (Pull Request, Push Message, etc)? May be worth just clarifying the "raw" numbers here

Reply:
Added a comment.

```rust
// bincode::serialize{,_into} serializes enum tag as a u32 (4 bytes)
// with fixint encoding in little endian.
```